spark/v3.4/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteMergeIntoTable.scala

```diff
@@ -214,6 +214,8 @@ object RewriteMergeIntoTable extends RewriteRowLevelIcebergCommand with Predicat
 
     val rowFromSourceAttr = resolveAttrRef(ROW_FROM_SOURCE_REF, joinPlan)
     val rowFromTargetAttr = resolveAttrRef(ROW_FROM_TARGET_REF, joinPlan)
+    // The output expression should retain read attributes for correctly determining nullability
+    val matchedOutputsWithAttrs = matchedActions.map(matchedActionOutput(_, metadataAttrs) :+ readAttrs)
```

Comment:
Can we reuse `matchedOutputs` above in this statement?

Reply:
This doesn't make sense to me. This does have the intended effect to correctly set the nullability in `buildMergeRowsOutput`, it seems to me like there is a better way than creating a different (fake) version of `matchedOutputs`. I think this works because it is including the original attributes as fake outputs (_for every action output list_). A more direct way of doing the same thing is to make an attribute nullable if it is nullable OR if an output is nullable.

In other words, I think the actual problem is in `RewriteRowLevelIcebergCommand.buildMergingOutput`:

```scala
  protected def buildMergingOutput(
      outputs: Seq[Seq[Expression]],
      attrs: Seq[Attribute]): Seq[Attribute] = {

    // build a correct nullability map for output attributes
    // an attribute is nullable if at least one output may produce null
    val nullabilityMap = attrs.indices.map { index =>
      index -> outputs.exists(output => output(index).nullable)
    }.toMap

    attrs.zipWithIndex.map { case (attr, index) =>
      AttributeReference(attr.name, attr.dataType, nullabilityMap(index))()
    }
  }
```

The change in the current PR updates the `outputs` arg to that method to include a copy of the original `attrs` for every matched expression. What ends up happening is that `outputs.exists` checks whether the corresponding original attr is nullable. You could do the same without this PR's change like this:

```scala
  protected def buildMergingOutput(
      outputs: Seq[Seq[Expression]],
      attrs: Seq[Attribute]): Seq[Attribute] = {

    // build a correct nullability map for output attributes
    // an attribute is nullable if at least one output may produce null
    val nullabilityMap = attrs.indices.map { index =>
      index -> outputs.exists(output => output(index).nullable)
    }.toMap

    attrs.zipWithIndex.map { case (attr, index) =>
      AttributeReference(attr.name, attr.dataType, attr.nullable || nullabilityMap(index))()
    }
  }
```

Looking at this code also raises a larger question: what is `buildMergingOutput` trying to do? It basically copies the input attrs to be the `MergeRows` output attrs, but then bases nullability on whether unrelated expressions are nullable. The problem is the expression coming from an output comes from an `Assignment` and has nothing to do with the position in the input attrs.

@aokolnychyi do you know what's happening here? Why would the output of assignment expressions affect the nullability of the incoming data?

Reply:
Okay, I think I understand why the inputs would be aligned with the outputs. This is the copy-on-write case, where each matched clause should produce an output for every column in order. If that's the case, then the indexing on each output should work. That is, each output should correspond to an input attr because both are based on the target table.

If that's the case, then I guess I can see what is happening here. The input and output correspond, so the input type and name should be used. But the output does depend on the nullability of the output expression.

If that's right, then the short-term fix is what I pasted above. If the table column is optional, then this should produce an optional output in case there is a null value.

That will work, but it doesn't explain why the output nullability is incorrect.

Reply:
Looks like the new tests pass locally with just the change to `buildMergingOutput`.

I also think that I can explain why the nullability is incorrect. That's because in the test case, the only expressions are constants (non-null) and there is no explicit case that represents the existing rows.

That means there are two possible solutions. First, the one I just tested where `buildMergingOutput` carries through the nullability of the input attrs. Assuming that the input attrs are nullable if and only if the table fields are nullable, that's probably the safest option. Second, the copy-on-write cases could pass an additional seq of the input attrs to represent the unmodified rows in the _unmatched_ set:

```scala
      // line 230
      output = buildMergeRowsOutput(matchedOutputs, notMatchedOutputs :+ readAttrs, readAttrs)
```

I think that I prefer the second option because the output of the `MergeRows` node is more accurate. If an output is always present then it will not be nullable. That's probably also safer.

I also validated that this passes tests.

Reply:
Thanks @rdblue appreciate the detailed thread with the reasoning, I followed along and it makes sense. I agree that adding the readAttrs to the notMatchedOutput makes more sense for having a `MergeRows` logical plan which is logically sound. I've updated the PR. @aokolnychyi let us know what you think.